### ISSUE TITLE
Understanding cache pruning

### ISSUE THREAD

**Issue by FedericoTrifoglio** [r-lib/cachem#10]
I'm trying to understand how pruning works. My goal is to use it for a shiny app on shinyapps.io. I'd like certain objects to be cached and shared by multiple users. Of course, I'm aware that shinyapps.io doesn't have persistent storage, but I'm happy with that as I actually want `max_age` not to be `Inf`. 24 hours is probably the maximum I would go with.

```
app_cache <- cachem::cache_disk(file.path(dirname(tempdir()), "my-app"), max_age = 10)
shinyOptions(cache = app_cache)
# my app here
```
With this, the rds file in `temp/my-app` is not actually deleted after 10 seconds. If I understand correctly pruning only occurs at least every 20 minutes and something needs to trigger `set()` another time before the file is deleted. 

So for example
12:00:00 - User 1 uses the app (`foo.rds` is cached)
12:00:05 - User 2 uses the app and benefits from caching
12:00:10 - `foo.rds` expires but still exists in `temp/my-app`
12:20:00 - User 3 uses the app (`set()` triggers `prune()` which deletes `foo.rds`), User 3 doesn't benefit from caching

Am I understanding this correctly?

**Comment by wch** [comment]
From the docs:

> Cache pruning occurs when `set()` is called, or it can be invoked manually by calling `prune()`.
>
> The disk cache will throttle the pruning so that it does not happen on every call to `set()`, because the filesystem operations for checking the status of files can be slow. Instead, it will prune once in every `prune_rate` calls to `set()`, or if at least 5 seconds have elapsed since the last prune occurred, whichever is first.


You are right that pruning does not happen by itself -- there is no background daemon that prunes after 10 seconds. However, I'm not sure where the 20 minutes in your question comes from.

The `prune_rate` defaults to 20. The name is not totally clear, but it is used to reduce the number of pruning operations if `set()` is called many times in rapid succession. If `set()` is called 80 times within 5 seconds, then pruning will only happen 4 times (1 in 20 calls to `set()`, where 20 is the `prune_rate`). However, if at least 5 seconds elapse between each call to `set()`, then pruning will happen each time.

In your example, you are mostly correct, although there is one detail that's not quite right. For user 3, when `get()` is called, it calls the internal function `maybe_prune_single_(key)` which checks if that individual key has expired, and if so, removes it from the cache. Then the `get()` results in a cache miss, and after that, the computation happens and `set()` is called.

https://github.com/r-lib/cachem/blob/596432048c779700bd904d9b2cef3ac86eb4b704/R/cache-disk.R#L262

**Comment by FedericoTrifoglio** [comment]
Many thanks Winston for replying and oh boy I feel so embarrassed now. I've misread `times` as `minutes` on the README. Was probably so fixated on the argument `max_age` that everything had to be measured in seconds or minutes. But yeah it makes total sense now.

**Comment by wch** [comment]
No worries. While I was answering your question, I noticed some bugs in the pruning throttling code, which I've fixed in #11.
